hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/helpers/KeyValueContainerUtil.java

```diff
@@ -73,14 +71,6 @@ public static void createContainerMetaData(File containerMetaDataPath, File
       throw new IOException("Unable to create directory for metadata storage." +
           " Path: " + containerMetaDataPath);
     }
-    MetadataStore store = MetadataStoreBuilder.newBuilder().setConf(conf)
-        .setCreateIfMissing(true).setDbFile(dbFile).build();
```

Comment:
Thanks for working on this @runzhiwang. Can we add to the cache here ?
Also the entry should be added to cache here and not in the apply transaction phase.

Reply:
@mukul1987 This is in HddsDispatcher.WriteData, if add to the cache here, 2 follower datanodes will not execute RocksDB.open in their HddsDispatcher.WriteData until the leader datanode finish HddsDispatcher.WriteData including RocksDB.open. So RocksDB.open in sequential.  If in apply transaction phase, RocksDB.open in parallel, it save 200ms. I think RocksDB was not used until HddsDispatcher.PutBlock, so add to cache in apply transaction phase does not matter. Please fix me if I am wrong.

Reply:
@runzhiwang , createContainer happens as a part of writeStateMachineData which gets executed on all nodes(leader and follower) independent of each other. I would prefer to add it to the container cache as soon as it gets created on the datanode on disk.

Reply:
@bshashikant Hi, but it looks like they are not independent of each other, follower must wait leader finish createContainer.

Reply:
> @bshashikant Hi, but it looks like they are not independent of each other, follower must wait leader finish createContainer.

Can we point me to the code where you see this happening or we can have a call to discuss this further offline?

Reply:
@bshashikant @mukul1987 
1. From the code,  the [future thread](https://github.com/apache/hadoop-ozone/blob/master/hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/transport/server/ratis/ContainerStateMachine.java#L672) do `getCachedStateMachineData ` in `readStateMachineData `and the [future thread](https://github.com/apache/hadoop-ozone/blob/master/hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/transport/server/ratis/ContainerStateMachine.java#L459) do `createContainer` in `writeStateMachineData `are the same [thread](https://github.com/apache/hadoop-ozone/blob/master/hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/transport/server/ratis/ContainerStateMachine.java#L505). Because `writeStateMachineData  `called before `readStateMachineData`. So leader must wait `createContainer `finish then `getCachedStateMachineData `and append logs to the follower, so leader and follower are not independent, follower must wait leader finish `createContainer`.  I think this order can be improved by distinguishing the thread used by `getCachedStateMachineData `  and `createContainer `. What do you think ?
2. From the jaeger UI, you can see follower create container after leader finishing it.
![image](https://user-images.githubusercontent.com/51938049/77489346-dbbac900-6e72-11ea-89e7-70b9f313e991.png)

Reply:
Thanks @runzhiwang . readStateMachine getting invoked after writeStateMachineData completes, is happening bcoz of recent change. i agree with ur proposal here.

Reply:
@bshashikant Thanks, I will open another PR to fix the dependent of leader and follower.
And do you think where we should put the RocksDB into cache ? Here or CommitData i.e. apply transaction phase. If place it here, once cache miss, follower must wait leader, the wait time including RocksDB.open which cost 200ms. If place it in CommitData, once cache miss, follower need not wait so long.